### PR TITLE
Update CFT tests and add them to the Jenkinsfile

### PR DESCRIPTION
**Pull request by knkski**
Allows running the script that controls the CFT tests locally, and adds them to the Jenkinsfile.

### CODE REVIEW

File: src/message_type.rs [hunk 1]
```diff
@@ -193,6 +193,23 @@ impl ParsedMessage {
 
         Self::from_peer_message(peer_message, true)
     }
+
+    /// Constructs a copy of this message with the given message type
+    #[allow(needless_pass_by_value)]
```
Comment (dcmiddle) [inline src/message_type.rs]: just curious, why is pass by value required here?

Comment (knkski) [reply src/message_type.rs]: The overall correct thing to do is going to be deriving `Copy` for `PbftMessageType`, see https://stackoverflow.com/a/44695401 for some rationale on that. However, making that change is a little out-of-scope for this PR, as it's going to involve changing a fair number of lines of code that are currently passing by reference. Passing by value is used here in anticipation of making that change, though.

Comment (dcmiddle) [reply src/message_type.rs]: thx

File: src/node.rs [hunk 2]
```diff
@@ -78,6 +78,8 @@ impl PbftNode {
         msg: ParsedMessage,
         state: &mut PbftState,
     ) -> Result<(), PbftError> {
+        info!("{}: Got peer message: {}", state, msg.info());
```
Comment (aludvik) [inline src/node.rs]: This commit should be reworded to explain what defects are being fixed. What is interesting to the reader is not that the test now passes, but what was broken before that you have fixed. This might need to be broken up into several commits to accomplish that, if there is more than one fix contained here.

Comment (knkski) [reply src/node.rs]: @aludvik: Does it suffice to say that these fixes are really just implementing node catchup part II due to rushing the previous code out the door for a deadline? They would've been squashed into the previous PR if there had been time.

Comment (aludvik) [reply src/node.rs]: I think it is important that the commit message be helpful in understanding what the commit is doing and why. "Fix CFT tests" gives me the impression the commit is modifying the tests. In reality it is making changes to the implementation in order to pass the tests, fixing a number of bugs in the process, so I think it is misleading.

To give you something more concrete, based on the commit description, I would suggest having one commit like this:
```
Synchronize sequence number on new block

Ensures that the sequence number doesn't get out of sync with actual state by
syncing them whenever a new block is received. This is required to pass the
liveness test.
```

and another like this:
```
Handle missing state after crash in catch up

Provides a more hardy version of catchup that is tolerant of booting up 
without any old state after a crash. This is required to pass the liveness test.
```

File: src/node.rs [hunk 3]
```diff
@@ -571,10 +571,12 @@ impl PbftNode {
             Ok(()) => {}
             Err(err) => {
                 warn!(
-                    "Proposing view change due to failed consensus seal verification! Error was {}",
+                    "Failing block due to failed consensus seal verification! Error was {}",
                     err
                 );
-                self.propose_view_change(state)?;
+                self.service.fail_block(block.block_id).map_err(|err| {
```
[the remark this answers is not in the document]
Comment (aludvik) [reply src/node.rs]: Fail block is necessary to inform the validator that the block is no longer being considered and uphold the contract that for every new block sent to the engine a commit, fail, or ignore will be sent back to the validator for the block.

I am also curious about why we no longer want to do a view change though.

[the remark this answers is not in the document]
Comment (knkski) [reply src/node.rs]: Good catch, now we've got both a block fail and view change.

File: src/node.rs [hunk 4]
```diff
@@ -1270,6 +1321,84 @@ mod tests {
         assert_eq!(state1.seq_num, 0);
     }
 
+    #[test]
+    fn block_new_first_10_blocks() {
```
Comment (knkski) [inline src/node.rs]: @aludvik, @ltseeley: In the interest of getting this merged I'm not going to add more unit tests beyond this to this PR, but it would be really useful to start writing some unit tests around various conditions such as catching up after having crashed without saving state.

File: src/node.rs [hunk 5]
```diff
@@ -1316,7 +1445,7 @@ mod tests {
             node.msg_log.add_message(message);
         }
 
-        let seal = node.build_seal(&mut state, vec![1, 2, 3], head).unwrap();
+        let seal = node.build_seal(&state, vec![1, 2, 3], head).unwrap();
```
Comment (ltseeley) [inline src/node.rs]: This change should be in the commit where you change the signature for this method

File: src/node.rs [hunk 6]
```diff
@@ -466,11 +466,19 @@ impl PbftNode {
...
+        info!(
+            "{}: Trying catchup from head #{} to #{}, from BlockNew message #{}",
+            state,
+            head.block_num,
+            head.block_num + 1,
+            block.block_num,
+        );
+
+        // Don't catch up from block 0 -> 1, since there's no consensus seal.
+        if block.block_num < 2 {
```
Comment (ltseeley) [inline src/node.rs]: Not returning early when the node is primary seems like a pretty significant change that isn't really captured by the commit message; this should probably be in its own commit.

File: src/node.rs [hunk 7]
```diff
@@ -483,14 +491,25 @@ impl PbftNode {
                 let block_id_matches = block.previous_id == wb.get_block_id();
 
                 if !block_num_matches || !block_id_matches {
+                    debug!(
```
Comment (ltseeley) [inline src/node.rs]: There are a lot of logging updates in this commit; perhaps these should be to moved to the "logging improvement" commit that's a part of this PR to keep this commit clean.

File: src/node.rs [hunk 8]
```diff
@@ -516,43 +535,51 @@ impl PbftNode {
                     Ok(msgs)
                 })?;
 
-        // Sync local state with actual state. We re-sync the sequence number with head
-        // again, even though we already did that in `on_block_new`, since if we're
-        // catching up as the primary, the sequence number should actually just be
-        // the head's block num, and not that+1.
-        if state.seq_num != head.block_num {
```
Comment (ltseeley) [inline src/node.rs]: This block of code was added in an earlier commit in this PR, so you should just fix it up with that commit to remove it altogether.

File: src/node.rs [hunk 9]
```diff
@@ -566,7 +593,27 @@ impl PbftNode {
...
+        info!(
+            "{}: Got BlockNew: {} / {}",
+            state,
+            block.block_num,
+            hex::encode(&block.block_id[..6]),
+        );
+
+        debug!("Head is currently at {}", head.block_num);
+
+        if block.block_num <= head.block_num {
```
Comment (ltseeley) [inline src/node.rs]: Again, this seems like a fairly self-contained change that isn't properly captured by the commit message. Probably better off with its own commit.

File: tests/test_crash_fault_tolerance.sh [hunk 10]
```diff
@@ -20,29 +20,51 @@
 # a node, verifying all nodes reach block 20, restarting the node, and
 # verifying that all nodes reach block 30.
 
-if [ -z "$ISOLATION_ID " ]; then export ISOLATION_ID=latest; fi
+if [ -z "$ISOLATION_ID" ]; then export ISOLATION_ID=latest; fi
```
Comment (ltseeley) [inline tests/test_crash_fault_tolerance.sh]: I don't think this commit is necessary, since we can already run the tests locally with `bin/run_docker_test` which is the recommended way of running it. Also, it makes the output of the test rather messy, and from my experience the containers aren't cleaned up properly when using `run_docker_test`

File: Jenkinsfile [hunk 11]
```diff
@@ -94,6 +94,23 @@ pipeline {
             }
         }
 
+        stage('Run CFT tests') {
+            options {
+                timeout(time: 10, unit: 'MINUTES')
+            }
+            steps {
+                sh 'tests/test_crash_fault_tolerance.sh'
```
Comment (ltseeley) [inline Jenkinsfile]: The Jenkins tests should be run using `bin/run_docker_test` (see https://github.com/hyperledger/sawtooth-raft/blob/master/Jenkinsfile)

File: src/node.rs [hunk 12]
```diff
@@ -516,43 +535,51 @@ impl PbftNode {
...
+        )?;
 
         // Start a view change if we need to force one for fairness
         if state.at_forced_view_change() {
             self.force_view_change(state);
         }
 
+        // Ensure that the message that we generated from the BlockNew message
+        // has the right sequence number.
+        let mut fixed_msg = msg.clone();
```
Comment (ltseeley) [inline src/node.rs]: Self-contained change

File: src/node.rs [hunk 13]
```diff
@@ -516,43 +535,51 @@ impl PbftNode {
...
+        fixed_msg.set_info(fixed_info);
+
         self.msg_log
-            .add_message(ParsedMessage::from_pbft_message(msg));
+            .add_message(ParsedMessage::from_pbft_message(fixed_msg));
         state.working_block = WorkingBlockOption::TentativeWorkingBlock(block.block_id.clone());
         state.idle_timeout.stop();
         state.commit_timeout.start();
 
+        if state.is_primary() {
```
Comment (ltseeley) [inline src/node.rs]: Self-contained change

File: src/node.rs [hunk 14]
```diff
@@ -516,43 +535,51 @@ impl PbftNode {
...
 
-        // Commit the new block, using one of the parsed commit messages to simulate
+        // Commit the new block, using one of the parsed messages to simulate
         // having received a regular commit message.
-        handlers::commit(state, &mut self.msg_log, &mut *self.service, &messages[0])?;
+        handlers::commit(
+            state,
+            &mut self.msg_log,
+            &mut *self.service,
+            &messages[0].as_msg_type(PbftMessageType::Commit),
```
Comment (ltseeley) [inline src/node.rs]: Another self-contained change that's not captured by the commit message.

File: tests/test_crash_fault_tolerance.sh (hunk 10 above)
Comment (knkski) [reply tests/test_crash_fault_tolerance.sh]: I really think we should move away from the docker-in-docker thing, I don't see the value-add for the complexity that it brings.

File: src/node.rs (hunk 12 above)
Comment (knkski) [reply src/node.rs]: There isn't really a single bug that each of these changes fix, so I don't see what splitting it up gives us - the commit message would just be something like "part of making catchup more robust".

File: tests/test_crash_fault_tolerance.sh (hunk 10 above)
Comment (ltseeley) [reply tests/test_crash_fault_tolerance.sh]: @aludvik could you weigh in on this?

File: src/node.rs (hunk 12 above)
Comment (ltseeley) [reply src/node.rs]: That's okay, 1 commit != 1 bug fix. Generally speaking, 1 PR should be 1 bug fix. Best practice from my experience is that the smallest self-contained change that results in runnable, improved code should be its own commit. Each of these changes is distinct; they are all in service of fixing the same bug, but they are distinct changes nonetheless. For this change, the comment you provided is already a pretty good commit message. In my opinion, it's a lot easier to understand what's going on when each of the changes is described, rather than all being lumped together and just saying "this makes it more robust."

File: tests/test_crash_fault_tolerance.sh (hunk 10 above)
Comment (aludvik) [reply tests/test_crash_fault_tolerance.sh]: @ltseeley I think this commit has a number of changes that are good, regardless of whether the test is ultimately run under docker or with run_docker_test. I would not say they are unnecessary and I don't think using `bin/run_docker_test` fixes the problems.

@knkski I would like to hear more about your opinions on why docker is bad and what you think some better alternatives are. I agree it can get complex, but the state of things is derived from a number of somewhat conflicting requirements on how we build up our integration tests such as:
1. Must be able to run the tests from any host, with a minimal set of requirements (usually docker, bash, and python).
2. Must be able to run the tests on Jenkins without separate jobs and tests interfering with each other.
3. Must not do docker-in-docker (this is why we mount the docker socket into the container, it is more of a docker-alongside-docker solution, I think this is the article I read when coming up with the solution: https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/)
